pexpect/expect.py

```diff
@@ -9,30 +9,41 @@ def __init__(self, spawn, searcher, searchwindowsize=-1):
         if searchwindowsize == -1:
             searchwindowsize = spawn.searchwindowsize
         self.searchwindowsize = searchwindowsize
-    
+
     def new_data(self, data):
         spawn = self.spawn
         searcher = self.searcher
 
-        incoming = spawn.buffer + data
-        freshlen = len(data)
-        index = searcher.search(incoming, freshlen, self.searchwindowsize)
+        pos = spawn._buffer.tell()
+        spawn._buffer.write(data)
+
+        # determine which chunk of data to search; if a windowsize is
+        # specified, this is the *new* data + the preceding <windowsize> bytes
+        if self.searchwindowsize:
+            spawn._buffer.seek(max(0, pos - self.searchwindowsize))
+            window = spawn._buffer.read(self.searchwindowsize + len(data))
+        else:
+            # otherwise, search the whole buffer (really slow for large datasets)
+            window = spawn.buffer
+        index = searcher.search(window)
         if index >= 0:
-            spawn.buffer = incoming[searcher.end:]
-            spawn.before = incoming[: searcher.start]
-            spawn.after = incoming[searcher.start: searcher.end]
+            value = spawn.buffer
+            spawn._buffer = spawn.buffer_type()
+            spawn._buffer.write(value[searcher.end:])
+            spawn.before = value[: searcher.start]
+            spawn.after = value[searcher.start: searcher.end]
             spawn.match = searcher.match
             spawn.match_index = index
             # Found a match
             return index
-    
-        spawn.buffer = incoming
-    
+        elif self.searchwindowsize:
+            spawn._buffer = spawn.buffer_type(window)
```

Comment:
I think this could do the wrong thing in some cases. I was just testing the BytesIO/StringIO constructors because I was going to suggest getting rid of the `.write()` on line 32, and I noticed that if you pass data into the constructor, it leaves the cursor at the beginning, so `.tell()` returns 0, and writing overwrites the data already there.

```python
import io
s = io.StringIO('abcd')
s.tell()  # -> 0
s.write('zz')
s.getvalue()  # -> 'zzcd'
```

Reply:
Which constructor?  I'm not sure I follow this comment.

Reply:
Oh, are you suggesting that I should instead do:

```python
spawn._buffer = spawn.buffer_type()
spawn._buffer.write(window)
```

?

Reply:
Yup, that's what I meant. That way, writing more data with `_buffer.write()` afterwards will add it to the end of the buffer, rather than overwriting existing data.